BJS2-40455 : Создание и настройка скиллов

lexsandr666 opened the pull request.

lexsandr666 committed:
BJS2-40455 : Создание и настройка скиллов

drewbal commented on src/main/java/school/faang/user_service/validation/skill/SkillDtoValidation.java:
```diff
@@ -0,0 +1,27 @@
+
+        if(skillDto.getTitle().length() > MAX_TITLE_LENGTH) {
+            throw new SkillDtoFieldConstraintValidationException("Наименование навыка превышает допустимое кол-во символов!");
+        }
```
А в исключения не надо цеплять значение, из-за которого оно вылетело? Или в логах будет видно?

drewbal commented:
Классный код, вопросов практически нет, подглядел у тебя интересные штуки, ну и еще завидую знанию неизвестных пока аннотаций :)

drewbal commented on src/main/java/school/faang/user_service/validation/skill/SkillDtoValidation.java:
```diff
@@ -0,0 +1,27 @@
+
+        if(skillDto.getTitle().length() > MAX_TITLE_LENGTH) {
+            throw new SkillDtoFieldConstraintValidationException("Наименование навыка превышает допустимое кол-во символов!");
+        }
```
А в исключения не надо цеплять значение, из-за которого оно вылетело? Или в логах будет видно?

lexsandr666 replied:
Если у тебя лог на километр не будет растягиваться, то можно =). Но в данном случае мы проверяем DTO на "пустышки", врятли в них что то есть.

RuslanZinovyev replied:
Лучше лог на километр чем без лога, уж поверь )
Кстати на забывай форматировать код у тебя if прилеплен к скобке

drewbal reviewed (commented):
Классный код, вопросов практически нет, подглядел у тебя интересные штуки, ну и еще завидую знанию неизвестных пока аннотаций :)
Остались тесты и to the moon!

lexsandr666 committed:
BJS2-40455 : Изменения.
1) Переделал Архитектуру SkillService
2) Переписал прямые обращения к UserRepository и SkillOfferRepository в классе SkillService на соотвествеющие классы сервисы (UserService и SkillService)
3) В процессе работы не мог поймать нормальыне Exception ввиду чего организовал GlobalExceptionHandler, которая прокидывал ошибку в Response
4) Создал как Unit, так и интеграционные тесты.

lexsandr666 committed:
Merge branch 'Dragon-master-stream7' into Dragon-master-stream7-bjs2-40455

lexsandr666 commented:
Вижу что есть ошибка в процессе тестов. При этом, локально выполняю тесты и они все проходят без ошибок.

<img width="313" alt="image" src="https://github.com/user-attachments/assets/53661f0d-727d-455a-bb51-293af385a2cf">

lexsandr666 committed:
BJS2-40455 : Fix Test Controller

lexsandr666 committed:
BJS2-40455 : Fix Test Controller

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,60 @@
+
+import java.util.List;
+
+@Component
```
Component аннотацию можно удалить, она дублирует аннотацию RestController

lexsandr666 replied:
Понял. Учту на будующее.

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,60 @@
+        return skillService.getOfferedSkills(skillCandidateDto);
+    }
+
+    @PostMapping("/acquireSkillFromOffers")
```
Мы не используем camel case для URL. хорошей практикой является использование kebab case
@PostMapping("/acquire-skill-from-offers")

Исправь также для остальных ендпоинтов

lexsandr666 replied:
Ок. Учту такую практику. Исправил у всех endpoint-ов содержажие несколько слов

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,60 @@
+public class SkillController {
+    private final SkillService skillService;
+    private final SkillDtoValidation skillDtoValidation;
+    private final SkillAcquireDtoValidation skillAcquireDtoValidation;
```
Здесь я могу тебе дать сразу 2 рекомендации.
Для начала советую убрать классы валидации SkillDtoValidation и SkillAcquireDtoValidation c контроллера и перенести их в сервис. На уровне контроллера у тебя уже есть аннотация Validated, просто добавь нужные правила валидации (constraints) на уровень DTO, а вся бизнес логика будет валидироваться уже в сервисе Skill Service. Также нет смысла создавать 2 класса валидации, создай один класс SkillValidation и объедини логику валидации в одном классе. Таким образом контроллер будет более легковеcным и чистым ровно с одной зависимостью от класса SkillService.

lexsandr666 replied:
1) Почему хотел проверять данные на стороне контроллера а не сервиса, что бы исключить лишний заход в метод, поэтому логику валидации dto вынес на сторону контроллера. Анотаций на валидацию полей более чем хватило, поэтому просто избавился от классов валидаторов DTO.
2) "Также нет смысла создавать 2 класса валидации" - тут пошёл от такой логики, что в будующем подобных DTO могло быть намного больше что вызвало бы проблема разростания одного файла до неодыкватных размеров, поэтому создал отдельный класс валидатор для каждой DTO.

Поправил.
Разобрался с анотациями и добавил их в классы (DTO). 
Убрал классы валидаторы за ненадобностью.
Протестировал через Postman.

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/skill/SkillAcquireDto.java:
```diff
@@ -0,0 +1,13 @@
+    private Long skillId;
+    private Long userId;
+
+    public SkillAcquireDto() {}
```
Если ты уже используешь библиотеку Lombok не проще ли добавить @NoArgsConstructor вместо создания пустого конструктора?

lexsandr666 replied:
Исправлено.

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/skill/SkillCandidateDto.java:
```diff
@@ -0,0 +1,14 @@
+    private SkillDto skill;
+    private Long offersAmount;
+
+    public SkillCandidateDto() {}
```
Если ты уже используешь библиотеку Lombok не проще ли добавить @NoArgsConstructor вместо создания пустого конструктора?

lexsandr666 replied:
Исправлено.

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,14 @@
+    private Long id;
+    private String title;
+
+    public SkillDto() {}
```
Если ты уже используешь библиотеку Lombok не проще ли добавить @NoArgsConstructor вместо создания пустого конструктора?

lexsandr666 replied:
Исправлено.

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/ValidateException.java:
```diff
@@ -0,0 +1,14 @@
+import school.faang.user_service.exception.skill.SkillDtoFieldConstraintValidationException;
+
+public class ValidateException extends RuntimeException {
+    private static final Logger logger = LoggerFactory.getLogger(SkillDtoFieldConstraintValidationException.class);
```
Если ты используешь библиотеку Lombok  в своем проекте ты можешь просто добавить аннотацию @Slf4j
На уровне класса и логгер c именем переменной log уже будет доступен. 
`log.error(message);`

lexsandr666 replied:
Понял. Учту. Поправил.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.stereotype.Component;
+import school.faang.user_service.repository.recommendation.SkillOfferRepository;
+
+@Component
```
Хорошей практикой является использование аннотации Service для сервиса вместо Component

lexsandr666 replied:
Понял. Учту. Поправил.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+
+import java.util.List;
+
+@Component
```
Ты уже достаточно продвинутый студент, поэтому рекомендую использовать Service вместо Component

lexsandr666 replied:
Поправил.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+    private final SkillValidation skillValidation;
+    private final UserService userService;
+    private final SkillOfferService skillOfferService;
+    public final int OFFERS_AMOUNT = 3;
```
Ты забыл добавить static и раз это константа она должна идти первой

lexsandr666 replied:
Да, статик всегда стараюсь на верх выкидывать.
Изначально предполагал что данное поле будет использоваться только внутри сервиса не "торча" наружу. В тесте понадобилось использовать поле и сделал его public, ввиду чего, логичнее было его еще и сделать static, так как оно все равно с модификатором доступа final

Поправил.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+        User user = userService.getByIdOrThrow(skillAcquireDto.getUserId());
+
+        if(user.getSkills().contains(skill)) {
+            throw new SkillDuplicateException("Пользователь " + user.getUsername() + " уже владеет навыком " + skill.getTitle());
```
Желательно любые ошибки писать на английском языке, я сомневаюсь что даже в РФ компании используют русский в разработке.

lexsandr666 replied:
Ну... за 5 лет практики, 3 продукта и что в БД и что в программах лог был весь русский, если конечно это не классический Exception или StackTrace. На текущей работе это основное требование к продукту, поэтому пишу на автомате =).

Поправил и учту на последующих логах и ошибках.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+        return skillMapper.toDto(skill);
+    }
+
+    public Skill getByIdOrThrow(Long id) {
```
getSkillByIdOrThrow более подходящее имя

lexsandr666 replied:
Поправил. Смущало в названии "OrThrow"

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+        Skill skill = skillMapper.toEntity(skillCandidateDto.getSkill());
+        int offersAmount = skillOfferService.getCountSkillOffersBySkill(skill.getId());
+
+        skillCandidateDto.setOffersAmount(Long.valueOf(offersAmount));
```
Чтобы избежать этой оберкти Long.valueOf(offersAmount)
тебе проще заставить getCountSkillOffersBySkill(skill.getId()) вернуть long

lexsandr666 replied:
Ну да, логично сразу из сервиса (внутри из репозитория) доставать поле Long. Совсем не подумал, поправил cервис и метод у репозитория сущности skillOffer.

Поправил.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public Skill getByIdOrThrow(Long id) {
+        Skill skill = skillRepository.getById(id);
```
```
return Optional.ofNullable(skillRepository.getById(id))
            .orElseThrow(() -> new SkillResourceNotFoundException("Skill not found in DB with id = " + id));
```
тогда проверка на null будет не нужна, если ты еще не работал с Optional это удобный класс чтобы избегать NPE

lexsandr666 replied:
К сожалению не успел на прошлых стримах должным образом поработать с Optional, поэтому просто из головы вылетез этот механизм. 

Поправил. Постараюсь наверстать упущенные инструменты.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -3,13 +3,23 @@
     private final UserRepository userRepository;
+  
+    public User getByIdOrThrow(Long id) {
+        User user = userRepository.getById(id) ;
```
Здесь можно использовать ту же логику с классом Optional что я указал выше

lexsandr666 replied:
Тут сделал проще, в доработке у Иры в сервисе уже есть метод findUser выполняющую логику получания пользователя с Exception. Удалил свой метод и переиспользовал её реализацию исключив "дублирующие логику" методы.

lexsandr666 committed:
BJS2-40455 : Cleaning up the code

lexsandr666 committed:
Merge branch 'Dragon-master-stream7' into Dragon-master-stream7-bjs2-40455

# Conflicts:
#	src/main/java/school/faang/user_service/controller/SkillController.java
#	src/main/java/school/faang/user_service/dto/skill/SkillCandidateDto.java
#	src/main/java/school/faang/user_service/dto/skill/SkillDto.java
#	src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java
#	src/main/java/school/faang/user_service/exception/skill/SkillDuplicateException.java
#	src/main/java/school/faang/user_service/service/SkillOfferService.java
#	src/main/java/school/faang/user_service/service/SkillService.java

lexsandr666 committed:
BJS2-40455 : Cleaning up the code

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
+
 @Validated
 @RestController
 @RequestMapping("/api/skills")  // Базовый URL для контроллера
```
Добавь, пожалуйста, версионирование /api/v1/skills

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
 
-    @PostMapping
-    public SkillDto create(@Valid @RequestBody SkillDto skillDto) {
+    @PostMapping("/create")
```
Я думаю, что можно не указывать ручку /create, поскольку ты и так дергаешь метод post. По спецификации RESTfull, по которой мы пишем контроллеры идет указание post метода для создания сущности. Поэтому указывать отдельно не обязательно. Проще для чтения и понимания.

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
-    @PostMapping
-    public SkillDto create(@Valid @RequestBody SkillDto skillDto) {
+    @PostMapping("/create")
+    public SkillDto create(@Valid @RequestBody SkillDto skillDto) throws Exception {
```
Ответ заверни в ResponseEntity<SkillDto>

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
-    @GetMapping("/users/{userId}")
-    public List<SkillDto> getUserSkills(@PathVariable @NotNull @Positive long userId) {
+    @GetMapping("/user/{userId}")
+    public List<SkillDto> getUserSkills(@PathVariable @Min(1) Long userId) throws Exception {
```
Так же заверни ответ в ResponseEntity

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
-    public List<SkillDto> getOfferedSkills(@PathVariable @NotNull@Positive long userId) {
-        return skillService.getOfferedSkills(userId);
+    @PostMapping("/offered-skills")
+    public SkillCandidateDto getOfferedSkills(@Valid @RequestBody SkillCandidateDto skillCandidateDto) throws Exception {
```
И тут заверни ответ в ResponseEntity

YTXIRE commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -11,30 +10,38 @@
+    }
+
+    @PostMapping("/acquire-skill-from-offers")
+    public SkillDto acquireSkillFromOffers(@Valid @RequestBody SkillAcquireDto skillAcquireDto) throws Exception {
```
И тут так же ResponseEntity

YTXIRE commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -2,11 +2,41 @@
+        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(Exception.class)
```
Лучше указывать конкретное исключение, нежели ловить группу.

YTXIRE commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -12,6 +12,14 @@
     private final SkillOfferRepository skillOfferRepository;
 
+    public int getCountSkillOffersForUser(Long skillId, Long userId) {
+        return skillOfferRepository.countAllOffersOfSkill(skillId, userId) ;
```
А где перегруженный метод?

YTXIRE commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -2,15 +2,116 @@
+            .toList();
+    }
+
+    @Transactional
```
Она тебе здесь сильно нужна?
Мне кажется у тебя тут одна операция происходит

YTXIRE commented on src/test/java/school/faang/user_service/controller/SkillControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+    public void createSkillWithoutTitle() throws Exception {
+        SkillDto invalidSkillDto = new SkillDto(null, null);
+
+        mockMvc.perform(post("/api/skills/create")
```
Тут так же версионирование

YTXIRE commented on src/test/java/school/faang/user_service/controller/SkillControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+        doNothing().when(skillValidation).validateDuplicate(any());
+        when(skillService.create(any(SkillDto.class))).thenReturn(skillDtoJava);
+
+        mockMvc.perform(post("/api/skills/create")
```
И тут версионирование

YTXIRE commented on src/test/java/school/faang/user_service/controller/SkillControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+
+        when(skillService.getUserSkills(1L)).thenReturn(skills);
+
+        mockMvc.perform(get("/api/skills/user/1"))
```
И тут версионирование
